File: MergeSort.cpp
procedure mergesort( array,N )
array – list of elements to be sorted
N – number of elements in the list
begin
if ( N == 1 ) return array
var array1 as array = a[0] ... a[N/2]
var array2 as array = a[N/2+1] ... a[N]
array1 = mergesort(array1)
array2 = mergesort(array2)
return merge( array1, array2 )
end procedure
procedure merge(array1, array2 )
    array1 – first array
    array2 – second array
begin
var c as array
while ( a and b have elements )
if ( array1[0] > array2[0] )
             add array2 [0] to the end of c
             remove array2 [0] from array2
else
             add array1 [0] to the end of c
             remove array1 [0] from array1
end if
end while
while ( a has elements )
add a[0] to the end of c
remove a[0] from a
end while
    
while ( b has elements )
add b[0] to the end of c
remove b[0] from b
end while
    
return c
end procedure

File: MergeSort.java
public class MergeSort {

    // Main method to sort an array using merge sort
    public static void mergeSort(int[] arr) {
        if (arr.length < 2) {
            return; // Base case: an array of length 0 or 1 is already sorted
        }

        // Split the array into two halves
        int mid = arr.length / 2;
        int[] left = new int[mid];
        int[] right = new int[arr.length - mid];

        // Fill the left and right arrays
        for (int i = 0; i < mid; i++) {
            left[i] = arr[i];
        }
        for (int i = mid; i < arr.length; i++) {
            right[i - mid] = arr[i];
        }

        // Recursively sort both halves
        mergeSort(left);
        mergeSort(right);

        // Merge the sorted halves
        merge(arr, left, right);
    }

    // Helper method to merge two sorted arrays
    private static void merge(int[] arr, int[] left, int[] right) {
        int i = 0, j = 0, k = 0;

        // Merge while there are elements in both arrays
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                arr[k++] = left[i++];
            } else {
                arr[k++] = right[j++];
            }
        }

        // Copy remaining elements of left array
        while (i < left.length) {
            arr[k++] = left[i++];
        }

        // Copy remaining elements of right array
        while (j < right.length) {
            arr[k++] = right[j++];
        }
    }

    // Main method to test the merge sort implementation
    public static void main(String[] args) {
        int[] arr = {38, 27, 43, 3, 9, 82, 10};

        System.out.println("Unsorted array:");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();

        mergeSort(arr); // Sort the array

        System.out.println("Sorted array:");
        for (int num : arr) {
            System.out.print(num + " ");
        }
    }
}
